### DISCUSSION TITLE
CF App Cutting Transfers Off

### DISCUSSION THREAD

**Discussion by stephen-hunt** [nasa/cFS#768] [Q&A] [answered]
Hello,

I've been running file transfers with the CF app and a ground CFDP Entity. Things are looking mostly good. That's until I tried doing a larger file transfer. After a certain size, I noticed the CF app stopped sending out PDUs. With some analysis, I've realized the CF app consistently stops sending after about 944 PDUs. This is across transfers as well. I.e. if I start one transfer with ~900 PDUs, and I start another transfer, that transfer will stop at ~44 PDUs.

I tried changing config parameters (Increased CF_NAK_MAX_SEGMENTS and CF_R2_CRC_CHUNK_SIZE), within cf_platform_cfg.h, but that didn't help.

Does anyone know of other config parameters that may help with this situation? Or is there a known bug?

Thank you,
Stephen

**Comment by skliper** [comment]
We regularly run scenarios where CF generates ~60,000 PDUs with no identified issues (class 1 via the polling directory, all on a single channel).  Looks like you might be using class 2 though?  Can you swap to class 1 and see if it works?

I'm not as familiar with class 2, but are you able to detail the flow of PDU types in each direction to confirm the ground entity is providing the expected responses?

**Reply by stephen-hunt** [reply]
Interesting. My issue is with the CF_TxFileCmd_t Command, Class1 and Class2 downlink. I'll try the poll directory command, I haven't tested that at all yet. 

I would assume with Class 1, there's not many responses needed from the ground software. Are there "periodic acks" or something needed from the ground entity to help with Class 1? I didn't think so.

**Reply by skliper** [reply]
Correct, no responses for Class 1.  Since you are seeing the issue on both class 1 and 2, are you doing anything for flow control (maybe the flow control semaphore)?  Maybe something to do with transport layer buffers or similar?

**Reply by skliper** [reply]
What version CF are you using?  We are pretty close to main (main as of March 14, 12eff1c97935942fb9aaa57f44c5ad50918b35b4).  If your version is fairly old might be worth skimming closed issues.

**Reply by stephen-hunt** [reply]
I'm using Draco-rc4. We don't have equuleus CFE compatibility. 

And it's not flow control. I'm running the command with a ~5.2 MB file and 186 outgoing_file_chunk_size (set it cf_def_config.c). I have printouts in CF_CFDP_SendFd() showing me that that the sends abruptly stop, even while the Engine continues to run. And I know I'm not sending all the PDUs, because I have counters which tell me how many PDUs are sent out, before they stop getting sent.

**Reply by skliper** [reply]
Anything interesting in HK, maybe in the channel info (hk faults or anything strange w/ queue sizes)?  Maybe you're running out of software bus shared memory?  Is an EOF sent with error information?  Any events?

**Reply by stephen-hunt** [reply]
I haven't looked too much at the HK yet. I can do that. And no, neither EOF or EOT are being sent with information.

**Reply by stephen-hunt** [reply]
I am thinking some kind of internal buffer, that's why I was curious about configurations. Right now I'll play around more with different commands and continue to debug and find exactly where things are stopping and why.

**Reply by stephen-hunt** [reply, marked as answer]
So, this was a me problem. Inspired by a discussion on the CF app, I made changes to remove the Software Bus from the app and use POSIX pipes instead. Those changes have been quite helpful to transfer robustness. 

But I created a bug where CFE_SB_ReleaseMessageBuffer(CF_AppData.engine.out.msg); was getting called when producing the output message, but CFE_SB_ReleaseMessageBuffer(CF_AppData.engine.out.msg); was never called. I assume the release also occurs in CFE_SB_TransmitBuffer(CF_AppData.engine.out.msg, true); 

Releasing the buffer after sending it into the pipe resolved the problem. 

Thank you for the responses and help.
